trillian/integration/copier.go

```diff
@@ -0,0 +1,211 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package integration
+
+import (
+	"context"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"math/rand"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/google/certificate-transparency-go/client"
+	"github.com/google/certificate-transparency-go/scanner"
+	"github.com/google/certificate-transparency-go/trillian/ctfe"
+	"github.com/google/certificate-transparency-go/trillian/ctfe/configpb"
+	"github.com/google/certificate-transparency-go/x509"
+
+	ct "github.com/google/certificate-transparency-go"
+)
+
+// CopyChainGenerator creates certificate chains by copying suitable examples
+// from a source log.
+type CopyChainGenerator struct {
+	start, limit             time.Time
+	sourceRoots, targetRoots *ctfe.PEMCertPool
+	certs, precerts          chan []ct.ASN1Cert
+}
+
+// NewCopyChainGenerator builds a certificate chain generator that sources
+// chains from another source log, starting at startIndex (or a random index in
+// the current tree size if startIndex is -1).  This function starts background
```

Comment:
If you're going to mix args and sentinels then I think any -ve value should be used. Otherwise you can ask for startIndex -2 for example.

Reply:
Done.